src/providers/cryptoauthlib/aead.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright 2021 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+// CAL supports CCM with:
+// - tag lenght must be <4,16> and must be odd number
+// - nonce lenght must be <7,13>
+// CAL supports GCM with:
+// - tag lenght must be <12,16>
+// - nonce lenght must be <7,13>
+
+use super::Provider;
+use crate::authenticators::ApplicationName;
+use crate::key_info_managers::KeyTriple;
+use log::error;
+use parsec_interface::operations::psa_algorithm::{Aead, AeadWithDefaultLengthTag};
+use parsec_interface::operations::{psa_aead_decrypt, psa_aead_encrypt};
+use parsec_interface::requests::{ProviderId, ResponseStatus, Result};
+
+const DEFAULT_TAG_LENGTH: usize = 16;
+
+pub fn get_tag_length(alg: &Aead) -> Option<usize> {
+    match alg {
+        Aead::AeadWithDefaultLengthTag(AeadWithDefaultLengthTag::Ccm) => Some(DEFAULT_TAG_LENGTH),
+        Aead::AeadWithDefaultLengthTag(AeadWithDefaultLengthTag::Gcm) => Some(DEFAULT_TAG_LENGTH),
+        Aead::AeadWithShortenedTag {
+            aead_alg: AeadWithDefaultLengthTag::Ccm,
+            tag_length,
+        } => Some(*tag_length),
+        Aead::AeadWithShortenedTag {
+            aead_alg: AeadWithDefaultLengthTag::Gcm,
+            tag_length,
+        } => Some(*tag_length),
+        _ => None,
+    }
+}
+
+pub fn is_ccm_selected(alg: &Aead) -> bool {
+    matches!(
+        alg,
+        Aead::AeadWithDefaultLengthTag(AeadWithDefaultLengthTag::Ccm)
+            | Aead::AeadWithShortenedTag {
+                aead_alg: AeadWithDefaultLengthTag::Ccm,
+                ..
+            }
+    )
+}
+
+impl Provider {
+    pub(super) fn psa_aead_encrypt_internal(
+        &self,
+        app_name: ApplicationName,
+        op: psa_aead_encrypt::Operation,
+    ) -> Result<psa_aead_encrypt::Result> {
+        match get_tag_length(&op.alg) {
+            Some(tag_length) => {
+                let key_triple =
+                    KeyTriple::new(app_name, ProviderId::CryptoAuthLib, op.key_name.clone());
+                let key_id = self.key_info_store.get_key_id::<u8>(&key_triple)?;
+                let key_attributes = self.key_info_store.get_key_attributes(&key_triple)?;
+                op.validate(key_attributes)?;
+
+                let aead_param_gcm = rust_cryptoauthlib::AeadParam {
+                    nonce: op.nonce.to_vec(),
+                    tag_length: Some(tag_length as u8),
+                    additional_data: Some(op.additional_data.to_vec()),
+                    ..Default::default()
+                };
+
+                let aead_param_ccc = aead_param_gcm.clone();
```

Comment:
Maybe this line can be put inside the `if` line 70 to only do it when required? Same for decrypt